core/src/triggers.rs

```diff
@@ -40,24 +40,27 @@ impl TriggerSet {
     /// # Errors
     /// - If [`TriggerSet`] already contains a trigger with the same [`EventFilter`].
     /// It's the user's responsibility to first `Unregister` the `Trigger`.
-    pub fn add(&self, trigger: Trigger) -> Result<(), smartcontracts::Error> {
-        let action = trigger.action;
-        if self.hooks.contains_key(&trigger.id) {
-            Err(smartcontracts::Error::Repetition(
+    pub fn add(&self, trigger: &Trigger) -> Result<(), smartcontracts::Error> {
+        let action = trigger.action();
+
+        if self.hooks.contains_key(trigger.id()) {
+            return Err(smartcontracts::Error::Repetition(
                 InstructionType::Register,
-                IdBox::TriggerId(trigger.id),
-            ))
-        } else {
-            self.hooks.insert(trigger.id.clone(), action).map_or_else(
+                IdBox::TriggerId(trigger.id().clone()),
+            ));
+        }
+
+        self.hooks
+            .insert(trigger.id().clone(), action.clone())
```

Comment:
ah, since internal representation of data model structures isn't known from the perspective of `iroha_core` it's no longer possible to destructure the given object such as `Trigger`. Unless, I introduce methods which take by value. I'm not so sure how to proceed with this

Reply:
it might not be such a bad idea to expose by value getters

Reply:
there is additional problem here. Let me demonstrate with an example:
```rs
struct Kita {
    a: String,
    b: String,
}

impl Kita {
    fn take_a(self) -> String {
        self.a
    }
    fn take_b(self) -> String {
        self.b
    }
}
fn main() {
    let s = Kita {
        a: String::new(),
        b: String::new(),
    };
    s.take_a();
    s.take_b();		// Use after move!!!
```

This code doesn't compile because there is a use after move.

Reply:
Ineed that's a deep question. What about this *"hack"*:

```rust
struct Kita {
    a: String,
    b: String,
}

impl Kita {
    fn take_a(&mut self) -> String {
        std::mem::take(&mut self.a) // Leaves default in `self.a` and returns pervious value
    }
    fn take_b(&mut self) -> String {
        std::mem::take(&mut self.b)
    }
}
fn main() {
    let s = Kita {
        a: String::new(),
        b: String::new(),
    };
    s.take_a();
    s.take_b();
}
```

Reply:
Make `Action` implement `Default` can be a problem... Then trigger can have `Option<Action>` instead of just `Action` but it's also not a very good idea

Reply:
@Arjentix I'd leave trigger fields public for now